docs/man/lyrictag.1
.TH lyrictag 1 "January 16th, 2003"
.SH NAME
lyrictag \- graphically enter a step pattern
.SH SYNOPSIS
\fBlyrictag \-s \fIsongfile\fB \-l \fIlyricfile\fR [\fB\-o \fIoutput\fR] | [\fIfilename\fR]
.SH DESCRIPTION
This script will play an MP3 or Ogg file while it displays text from
a file, containing the lyrics of the song playing. As the lyric
comes up, press enter or space when it should be displayed during the song.
If you make a mistake and need to go back in the display, backspace will
go back one lyric. However, there is no way to go back in the song.
.PP
The 'a' key switches between outputting in terms of "atsec" and "waits"
(the default is "atsec"). This is probably unimportant unless you intend
to do a lot of post-lyrictag editing. The "waits" mode makes it easier to
copy and paste the resulting text (for e.g. refrains).
.PP
If, in the lyrics file, a line is enclosed with parentheses ( ( and ) )
then it will be considered a "translated" lyric, meaning it will appear
below regular lyrics, and in a different color.
.PP
.SH USAGE
.IP \-\-song\, \-s
Specify the song file to play when displaying the lyrics. This should
be MPEG Layer 3 or Ogg Vorbis (strongly preferred) format.
.IP \-\-lyrics,\ \-l
Read song lyrics to display from this file.
.IP \-\-output,\ \-o
Specify the file to output the lyrics to. If this is a pyDDR step file,
the lyrics will be appended if none exist; otherwise, a new file will
be created. The default output file is \fIlyricfile\fR.lyric.
.PP
\fBlyrictag\fR may also be invoked with only one argument (\fIfilename\fR).
If this is the case, that argument will be used as the file to read lyrics
from, \fIfilename\fR.step will be the output file, and \fIfilename\fR.ogg
will be the song file.
.SH AUTHOR
Brendan Becker <tgz@orotech.net> is the primary author of pyDDR. Joe
Wreschnig <piman@debian.org> is the maintainer of pyDDR for Debian and the
author of this manual page, and also hacked extensively on the current
lyrictag.
.SH SEE ALSO
\fBpyddr\fR(6)


ChangeLog
pydance ChangeLog

legend:
+ new feature added or made an existing one better
- removed a feature, or something bad happened
* fixed a bug

----- - CURRENT
0.7.5 - That's the second largest ChangeLog I've ever seen!	[2003.XX.XX]
+ New options: "Wide" to add jumps, holds on/off.
+ 5 panel (Pump It Up) support (single, versus/unison, couple, and double).
 + Support for KSF, SM, and .dance files with 5 panel steps.
 - Actual PIU pads may or may not work; none were available for testing.
+ Versus, couple, and double 6 and 8 panel modes.
+ New game type selector, so you can do e.g. Double/Endless.
+ A document on the KSF file format has been added in docs/.
+ The .dance specification has been updated with new modes and corrections.
+ A new centered display for single player modes.
+ New theming code.
 + Brings back oldschool and bryan themes.
 + Autodetect what game modes a theme will work for.
+ New 32x32 arrow set for 8/9 panel modes, by xsdg.
* Support the Thrustmaster USB mats (probably), and X-Box controllers.
 + We suspect that pydance will run on GNU/Linux on an X-Box now.
* Correctly encode/decode KSF, SM, DWI, and .dance file metadata.
 * KSF requires the Python Koco codecs package.
* DWI parsing for braindead constructs like "2!(2".
* Arrows now always draw in order (earlier under later).
* Many crashes fixed:
 * Stops + holds works again.
 * Oni lifebar works again.
 * Obscure edge case in 2 player endless mode works correctly.
* Parsing for all files is now more informative and less error-prone.
- Assist mode is broken for non-4 panel modes, we need more sounds.
- any2dance removed, it won't work anymore with this version.
- In-game menu help removed, due to inaccuracies.

0.7.4 - In My Pants						[2003.07.28]
+ 6 panel ("solo") and 8 panel support.
+ Double (1 player, both sides) support.
 - 'totaljudgings' option no longer works (no one was using it anyway...)
* Song selector UI tweaks (faster animations, nicer folder opening).
* Respace the arrows to be closer together in single/versus.
* All 9 directions can always be done on the numeric pad for player 1.
 - azerty 2nd player start key changed to '+' on the numeric pad.
* "Escape" changed from left+start to start+select.
* Some more speed optimizations for slower processors.
* Some DWI parser fixes for bad files.
* 'usepsyco' option (on by default) to try and load the Psyco JIT compiler.
- 'oldschool' and 'bryan' were removed (temporarily, maybe).

0.7.3
* Skipped, because the Linux Game Tome called 0.7.2 0.7.3 by mistake.

0.7.2 -	Dilaudid, how I love thee                               [2003.07.14]
+ Boost, brake, left, right, mirror, shuffle, and random ("supershuffle").
+ "Suck" lifebar mode, you fail if you do too good.
+ "Drop" lifebar mode, goes down at the regular speed but never up.
+ Folders in the song selector (needs > 60 songs).
* Parser bug fixes (especially for SM files).
* pyDDR has been renamed to pydance.
	
0.7.1 - OUCH MY BLADDER                                         [2003.07.01]
+ Support for COUPLE mode.
 + And battle mode.
 - any2dance doesn't always convert COUPLE from DWI to .dance correctly.
+ New player options screen (hold down start in the song selector).
+ New game options screen (hold down select).
+ "Ready? Go!" screen before the song starts.
+ New "oldschool" theme similar to earlier mixes.
* Standard round of bug fixes and tweaks.

0.7.0 - Applesauce makes the apples go yonder                   [2003.06.04]
+ More optimizations from JZig (movies this time)
+ New file formats
  + Native DWI and SM support
  + ".dance" instead of ".step" is now pyDDR's official format.
  + 'any2dance.py' converter for files.
* Judge rewrite!
  * Proper speed change handling in Judge
  * Proper miss handling in Judge

0.6.5 - Boy, did I mess that up.				[2003.05.30]
+ Major graphics optimization patch (1.5x-2x increase in fps) from JZig.
* Two player mode works again.

0.6.4 - RTFM. RTFM. RTFM. That means Read the Fucking Manual.	[2003.05.29]
+ New step file parsers
  + Beginner and hardcore (smaniac/oni/challenge) difficulty.
+ New hold arrow display
- 'mixerclock' option removed (it's now the only timing method).
- 'transcolor' option removed in favor of new 'lyriccolor' syntax.
* Bug fixes, optimizations, whippings, etc.
  + Most notably, holds, bpm changes, and stops are working much nicer now.
    - But may require reconversion with dwi2step.

0.6.3 - Reasonable Starting Times!				[2003.05.17]
+ New song loading display and algorithm (at least 4-5 times faster).
+ New song selector based roughly on DDR 5th Mix.
  + Nonstop help and playlist display.
  + 'gratuitous' option to turn off the slow extras in the song selector.
+ Endless mode.
+ Configurable oni life meter.
+ New "centered" arrow scrolling style for gameplay.
+ Smaller download size by converting wavs to oggs.
  - Requires SDL_mixer 1.2.5.
+ The timer display is controllable via the 'fpsdisplay' option.
+ New theme ('blue') and a new life display for all themes.
* DDR Extreme (8th Mix) grading and scoring algorithm.
* Split off the ddrmat.o driver from the main package.
* The constant cleanup/speedup/refactor/bug fix process.

0.6.2 - No More Ugly-Ass Buttons				[2003.04.28]
- Fullscreen switching during the game doesn't work on MacOS X.
  - Use the configuration file instead, for now.
- The pN_keys option is gone in favor of 'keyboard', and the 'sortpersist'
  option was removed.
+ 'killsongonfail' was renamed to 'autofail' and made default.
+ New game menu display.
+ Better theme detection and configuration support.
+ The 'badknees' option will remove jumps.
+ A new README that actually documents all the options!
* Once again, numerous cleanups and bugs fixed.

0.6.1 - The second one						[2003.04.09]
+ Experimental install system
  + Currently only works on Windows (and doesn't do much) and UNIX.
  + Manual pages included with pyDDR now.
  + New directory layout
+ Major code cleanups and bug fixes.

utils/lyrictag.py
#!/usr/bin/env python
# Rewritten version of a much hacked version of findbpm
# This program adds lyrics to stepfiles or other files

import sys
import time
import getopt
import os
import pygame, pygame.font, pygame.image, pygame.mixer
from pygame.locals import *

VERSION = "0.5"

def usage():
  print "lyrictag " + VERSION + " - time lyrics for PyDDR step files"
  print "Usage: " + sys.argv[0] + " [options]"
  print """\
  --song, -s      Music to play (Ogg Vorbis or MP3)
  --lyrics, -l    A text file containing lyrics to read from
  --output, -o    Filename for lyric data output

The lyrics in the file will be displayed, and one will be highlighted.
When the highlighted lyric is said in the song, press enter or space to
move the pointer to the next line. Backspace moves back a line.

If the output file is a PyDDR step file, the lyrics will be added to it
automatically. If no output file is given, the name of the lyric file will
have '.lyric' appended to it."""

def main():

  try:
    opts, args = getopt.getopt(sys.argv[1:], "s:l:o:hv",
                               ["song", "lyrics", "output", "help", "version"])
  except getopt.GetoptError:
    usage()
    sys.exit(2)

  lyric_file = None
  output_file = None
  song_file = None

  outfile = None

  for o, a in opts:
    if o in ("-h", "--help"):
      usage()
      sys.exit()
    elif o in ("-v", "--version"):
      print "lyrictag version " + VERSION
      sys.exit()
    elif o in ("-s", "--song"): song_file = a
    elif o in ("-l", "--lyrics"): lyric_file = a
    elif o in ("-o", "--output"): output_file = a

  if len(args) == 1:
    lyric_file = args[0]
    output_file = args[0] + ".step"
    song_file = args[0] + ".ogg"

  if song_file == None:
    if len(args) == 2: song_file = args[0]
    else: print "No song file given! Please use the --song (or -s) option"

  if lyric_file == None:
    if len(args) == 2: lyric_file = args[1]
    else: print "No lyric file given! Please use the --lyrics (or -l) option"

  if output_file == None and lyric_file != None:
    output_file = lyric_file + ".lyric"

  if not (output_file and song_file and lyric_file):
    print
    usage()
    sys.exit(3)

  text_file = open(lyric_file,"r")

  if os.access(output_file, os.W_OK) and output_file[-5:].lower() == ".step":
    outfile = open(output_file, "r+")
    for line in outfile.readlines():
      if line == "LYRICS\n":
        print "error: this step file already has lyrics"
        sys.exit(5)

  elif not os.access(output_file, os.F_OK):
    outfile = open(output_file, "w")
  elif not output_file[-6:].lower() == ".lyric":
    print "error: output file exists and is not a step file or lyric file"
    sys.exit(4)
  else:
    outfile = open(output_file, "w")
    
  outfile.write("LYRICS\n")

  lyric_text = []
  text_to_write = []
  fileline = '-'
  while fileline != '':
    fileline = text_file.readline()
    if fileline != '\n' and fileline != '':
      lyric_text.append(fileline[0:-1])
  lyric_text.append('[END OF FILE]')
 
  pygame.init()
  screen = pygame.display.set_mode((640, 480), HWSURFACE|DOUBLEBUF)
  pygame.display.set_caption('Lyric Tagger')
  screen.fill((0,0,0))

  pygame.mixer.music.load(song_file)
  pygame.mixer.music.play()
  beatstart = pygame.time.get_ticks()

  output_mode = 0 # 0 = atsec, 1 = waits
  current_time = 0
  last_time = 0

  current_line = 0
  total_lines = len(lyric_text) - 1

  font = pygame.font.Font(None, 48)

  while 1:
    time.sleep(0.0001)
    event = pygame.event.poll()

    screen.fill((0,0,0))
    for j in range(11):
      i = j - 5 
      c = i + current_line
      if c >= 0 and c <= total_lines:
        if j == 5:
          text = font.render(lyric_text[c], 1, (224, 224, 224))
        else:
          text = font.render(lyric_text[c], 1, (160, 160, 160))
        textpos = text.get_rect()
        textpos.centerx = screen.get_rect().centerx
        textpos.centery = 20 + (48*j)
        screen.blit(text, textpos)
      # show name
    pygame.display.flip()

    if event.type == QUIT or event.type == KEYDOWN and event.key == K_ESCAPE:
      break

    if not pygame.mixer.music.get_busy():
      break

    if event.type == KEYDOWN:
      if event.key == K_BACKSPACE and current_line != 0:
        current_line -= 1
        text_to_write.pop()
        text_to_write.pop()
      elif event.key == K_a:
        output_mode ^= 1
      elif current_line < total_lines:
        last_time = current_time
        current_time = pygame.time.get_ticks()/1000.0
        if output_mode == 0:
          text_to_write.append("atsec "+repr(current_time)+"\n")
        else:
          text_to_write.append("waits "+repr(current_time - last_time)+"\n")
          
        if (lyric_text[current_line][0] == "(" and
            lyric_text[current_line][-1] == ")"):
          text_to_write.append("trans "+lyric_text[current_line][1:-1]+"\n")
        else:
          text_to_write.append("lyric "+lyric_text[current_line] + "\n")
        current_line += 1

  for text in text_to_write:
    outfile.write(text)
  outfile.write("end\n")
  print "Song ended."

if __name__ == '__main__':
    main()
